src/sources/transient.rs

```diff
@@ -93,8 +96,43 @@
 /// The `TransientSource` will take care of updating the registration of the
 /// inner source, even if it actually needs to be unregistered or initially
 /// registered.
+///
+/// ## Replacing or removing `TransientSource`s without leaking
+///
+/// It is possible to leak registration if you bypass the API of a
+/// `TransientSource`. "Leak registration" means you may end up with an entry in
+/// `epoll()` that cannot be removed (in the case of file descriptor-based
+/// sources), or an entry in some other data structure (eg. the `Timer`'s
+/// underlying heap structure). No unsoundness or undefined behaviour will
+/// result, but leaking file descriptors can result in errors or panics in a
+/// long running program.
```

Comment:
Actually, I don't think it is that bad:

- If when dropping the event source you close the FD, when afaik epoll/kqueue will handle that gracefully, they don't internally keep it open and it won't cause errors.
- Leaking timers just means that the event loop will have a spurious wakeup at the time the timer expires.

Writing this, I'm actually wondering if there is actually much to worry about in the first place. :thinking:

Reply:
> If when dropping the event source you close the FD, when afaik epoll/kqueue will handle that gracefully, they don't internally keep it open and it won't cause errors.

I don't know about this. I checked this on an older version of Calloop (0.9.3) because that's what our code base was on. I added some logging statements for FDs getting added and removed from `epoll()`, and without these changes, the FDs just gradually increase as the edge case in the test is hit.

I **did not** do the same for timers, but it seems like the [binary heap in `TimerWheel`](https://github.com/Smithay/calloop/blob/613c839be75ba4f172d089bff0a94e5d3dfa3232/src/sources/timer.rs#L184) would just grow without actually unregistering the timers, even if they're dropped, no?

Reply:
> I don't know about this. I checked this on an older version of Calloop (0.9.3) because that's what our code base was on. I added some logging statements for FDs getting added and removed from epoll(), and without these changes, the FDs just gradually increase as the edge case in the test is hit.

I was pretty sure of the opposite, guess I was misremembering. You did close the FDs on dropping the sources?

> I did not do the same for timers, but it seems like the [binary heap in TimerWheel](https://github.com/Smithay/calloop/blob/613c839be75ba4f172d089bff0a94e5d3dfa3232/src/sources/timer.rs#L184) would just grow without actually unregistering the timers, even if they're dropped, no?

Yeah, but the timers would still fire and be consumed when they expire, it's just that no event source will actually react to them because their tokens no longer match anything. So unless you create a shit ton of timers into the far future and drop them without unregistering them, seems those would actually be short-lived leaks.

Reply:
I mean, the epoll man page states this pretty clearly imo:

> Will closing a file descriptor cause it to be removed from all epoll interest lists?
>
> Yes, but be aware of the following point.  A file descriptor is a reference to an open file description (see open(2)). Whenever a file descriptor is duplicated via dup(2), dup2(2), fcntl(2) F_DUPFD, or fork(2), a new file descriptor referring to the same open file description is created. An open file description continues to exist until all file descriptors referring to it have been closed.
> 
> A file descriptor is removed from an interest list only after all the file descriptors referring to the underlying open file description have been closed.  This means that even after a file descriptor that is  part  of  an interest  list  has been closed, events may be reported for that file descriptor if other file descriptors referring to the same underlying file description remain open.  To prevent this happening, the file descriptor must be explicitly removed from the interest list (using epoll_ctl(2) EPOLL_CTL_DEL) before it is duplicated.  Alternatively, the application must ensure that all file descriptors are closed (which may be difficult if file  descriptors were duplicated behind the scenes by library functions that used dup(2) or fork(2)).

Reply:
> You did close the FDs on dropping the sources?

These are all executors or channels, which should handle that internally, no?

> Yeah, but the timers would still fire and be consumed when they expire, it's just that no event source will actually react to them because their tokens no longer match anything.

That's a relief then.

I will see if I can repro this on 0.10.x. The fact that the fds increased suggest that something is holding them open, somehow. If nothing else though, even if `epoll()` removes fds from its own side, the fds-as-indexes will remain on "our" side in the `sys::Poll` `tokens` list, won't they?